### PR TITLE
Expansion proposal of `diffusers-cli env`

### PR DESCRIPTION
**Pull request by standardAI**
Coming from this discussion: #7345.
Do you want me to add more or remove some of them? Also, I thought of adding GPU's model if available, but I guess it violates privacy, right?

<details>
  <summary>transformers-cli env</summary>

  ```sh
  An NVIDIA GPU may be present on this machine, but a CUDA-enabled jaxlib is not installed. Falling back to CPU.

  Copy-and-paste the text below in your GitHub issue and FILL OUT the two last points.

  - `transformers` version: 4.38.2
  - Platform: Linux-6.5.0-26-generic-x86_64-with-glibc2.35
  - Python version: 3.10.12
  - Huggingface_hub version: 0.21.4
  - Safetensors version: 0.4.2
  - Accelerate version: 0.28.0
  - Accelerate config:  not found
  - PyTorch version (GPU?): 2.2.1+cu121 (True)
  - Tensorflow version (GPU?): not installed (NA)
  - Flax version (CPU?/GPU?/TPU?): 0.8.2 (CPU)
  - Jax version: 0.4.25
  - JaxLib version: 0.4.25
  - Using GPU in script?: <fill in>
  - Using distributed or parallel set-up in script?: <fill in>
  ```
</details>

<details>
  <summary>huggingface-cli env</summary>

  ```sh
  Copy-and-paste the text below in your GitHub issue.

- huggingface_hub version: 0.21.4
- Platform: Linux-6.5.0-26-generic-x86_64-with-glibc2.35
- Python version: 3.10.12
- Running in iPython ?: No
- Running in notebook ?: No
- Running in Google Colab ?: No
- Token path ?: /home/user_name/.cache/huggingface/token
- Has saved token ?: False
- Configured git credential helpers: cache
- FastAI: N/A
- Tensorflow: N/A
- Torch: 2.2.1
- Jinja2: 3.1.3
- Graphviz: N/A
- Pydot: N/A
- Pillow: 10.2.0
- hf_transfer: N/A
- gradio: N/A
- tensorboard: N/A
- numpy: 1.26.4
- pydantic: N/A
- aiohttp: 3.9.3
- ENDPOINT: https://huggingface.co
- HF_HUB_CACHE: /home/user_name/.cache/huggingface/hub
- HF_ASSETS_CACHE: /home/user_name/.cache/huggingface/assets
- HF_TOKEN_PATH: /home/user_name/.cache/huggingface/token
- HF_HUB_OFFLINE: False
- HF_HUB_DISABLE_TELEMETRY: False
- HF_HUB_DISABLE_PROGRESS_BARS: None
- HF_HUB_DISABLE_SYMLINKS_WARNING: False
- HF_HUB_DISABLE_EXPERIMENTAL_WARNING: False
- HF_HUB_DISABLE_IMPLICIT_TOKEN: False
- HF_HUB_ENABLE_HF_TRANSFER: False
- HF_HUB_ETAG_TIMEOUT: 10
- HF_HUB_DOWNLOAD_TIMEOUT: 10
  ```
</details>

@sayakpaul @yiyixuxu @DN6 @asomoza

### CODE REVIEW

File: src/diffusers/utils/import_utils.py [hunk 1]
```diff
@@ -295,6 +295,20 @@
 except importlib_metadata.PackageNotFoundError:
     _torchvision_available = False
 
+# Taken from https://stackoverflow.com/a/39662359.
+# Adapted to make it work with Google colab as well.
+_is_notebook = False
+try:
+    shell_class = get_ipython().__class__  # type: ignore # noqa: F821
+    for parent_class in shell_class.__mro__:  # e.g. "is subclass of"
+        if parent_class.__name__ == "ZMQInteractiveShell":
```
Comment (sayakpaul) [inline src/diffusers/utils/import_utils.py:304]: Is this universal enough?

File: src/diffusers/commands/env.py [hunk 2]
```diff
@@ -12,29 +12,64 @@
...
     def run(self):
         hub_version = huggingface_hub.__version__
 
+        safetensors_version = "not installed"
+        if is_safetensors_available():
+            import safetensors
+
+            safetensors_version = safetensors.__version__
+        elif util.find_spec("safetensors") is not None:
+            import safetensors
```
Comment (sayakpaul) [inline src/diffusers/commands/env.py:69]: Why is this needed?

File: src/diffusers/commands/env.py [hunk 3]
```diff
@@ -12,29 +12,64 @@
...
+
+            safetensors_version = f"{safetensors.__version__} but is ignored because of PyTorch version too old."
```
Comment (sayakpaul) [inline src/diffusers/commands/env.py:71]: The `else` of this should be `safetensors_version = "not installed"

File: src/diffusers/commands/env.py [hunk 4]
```diff
@@ -43,32 +78,75 @@ def run(self):
...
             "PyTorch version (GPU?)": f"{pt_version} ({pt_cuda_available})",
+            "Flax version (CPU?/GPU?/TPU?)": f"{flax_version} ({jax_backend})",
+            "Jax version": f"{jax_version}",
+            "JaxLib version": f"{jaxlib_version}",
             "Huggingface_hub version": hub_version,
             "Transformers version": transformers_version,
             "Accelerate version": accelerate_version,
+            "Accelerate config": f"{accelerate_config_str}",
+            "PEFT version": peft_version,
+            "SafeTensors version": f"{safetensors_version}",
```
Comment (sayakpaul) [inline src/diffusers/commands/env.py:149]:
```suggestion
            "Safetensors version": f"{safetensors_version}",
```

**Review by sayakpaul — COMMENTED** [review]
This is looking pretty tight. I am happy with the changes.

File: src/diffusers/commands/env.py (hunk 2 above)
Comment (standardAI) [reply src/diffusers/commands/env.py:69]: I forgot to mention. I took this directly from [`transformers`](https://github.com/huggingface/transformers/blob/243d0de9971d953b2a69636fba0420fb56cd42e4/src/transformers/commands/env.py#L62-L65).

File: src/diffusers/utils/import_utils.py (hunk 1 above)
Comment (standardAI) [reply src/diffusers/utils/import_utils.py:304]: I also forgot to mention this too. Taken from [`huggingface_hub`](https://github.com/huggingface/huggingface_hub/blob/57ee5fc97c20b884d97bac5e589204f6c4fb8c1b/src/huggingface_hub/utils/_runtime.py#L251-L264).

File: src/diffusers/commands/env.py (hunk 4 above)
Comment (standardAI) [reply src/diffusers/commands/env.py:149]: Copilot preferred this, I didn't notice :).

**Comment by standardAI** [comment]
Should I also add GPU's model if available due to this kind of situation: #2153.

**Comment by asomoza** [comment]
> Should I also add GPU's model if available to detect this kind of situation: https://github.com/huggingface/diffusers/issues/2153. However, I am unsure if this creates potential privacy issues.

What will help a lot is the amount of VRAM and maybe the architecture, but the VRAM would be good to know for the OOM issues.

I don't think it has privacy issues since the user have to willingly run the command and then post the information themselves, they can even edit the parts they don't want to share.

File: src/diffusers/commands/env.py (hunk 2 above)
Comment (sayakpaul) [reply src/diffusers/commands/env.py:68]: That's fine. It's okay to make things leaner here 😉

**Comment by sayakpaul** [comment]
> Should I also add GPU's model if available to detect this kind of situation: https://github.com/huggingface/diffusers/issues/2153. However, I am unsure if this creates potential privacy issues.

Agree with @asomoza's https://github.com/huggingface/diffusers/pull/7403#issuecomment-2011821072 comment :)

File: src/diffusers/commands/env.py
Comment (standardAI) [inline src/diffusers/commands/env.py:1]: Is `- Using GPU in script?: <fill in>` necessary now?

**Comment by standardAI** [comment]
I am still testing Windows part. Could anybody test Mac OS part if possible?

File: src/diffusers/commands/env.py
Comment (sayakpaul) [reply src/diffusers/commands/env.py:1]: Won't remove it.

**Comment by sayakpaul** [comment]
Tested on my Mac M2:

```bash
GPU Model: Apple M2 Max

Copy-and-paste the text below in your GitHub issue and FILL OUT the two last points.

- `diffusers` version: 0.28.0.dev0
- Platform: macOS-13.3-arm64-arm-64bit
- Running on a notebook?: No
- Running on Google Colab?: No
- Python version: 3.9.17
- PyTorch version (GPU?): 2.2.1 (False)
- Flax version (CPU?/GPU?/TPU?): 0.7.0 (cpu)
- Jax version: 0.4.13
- JaxLib version: 0.4.13
- Huggingface_hub version: 0.20.3
- Transformers version: 4.36.2
- Accelerate version: 0.25.0
- Accelerate config: 	not found
- PEFT version: 0.9.1.dev0
- Safetensors version: 0.3.1
- xFormers version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>
```

Could it make sense to make `GPU Model: Apple M2 Max` a part of the bullets? And call it "Accelerator"?

**Comment by standardAI** [comment]
I handled the merge conflict. Now, it is ready for a review.

**Comment by standardAI** [comment]
I handled the merge conflict. Now, it is ready for a review.
@sayakpaul @yiyixuxu @asomoza

File: src/diffusers/commands/env.py [hunk 5]
```diff
@@ -43,33 +74,116 @@ def run(self):
...
+                    stdout=subprocess.PIPE,
+                    stderr=subprocess.PIPE,
+                )
+                out_str, _ = sp.communicate()
+                out_str = out_str.decode("utf-8")
+
+                if len(out_str) > 0:
+                    accelerator = out_str.strip() + " VRAM"
+            except FileNotFoundError:
+                pass
```
Comment (sayakpaul) [inline src/diffusers/commands/env.py:144]: Could we maybe delegate the logic of finding accelerator specific stuff into separate utilities? This might be useful if we add support for a new accelerator.

**Review by sayakpaul — APPROVED** [review]
The changes look clean to me overall. This should be very helpful in the issues. Thanks a lot!

**Comment by sayakpaul** [comment]
Would like to @BenjaminBossan's thoughts too.

File: src/diffusers/commands/env.py [hunk 6]
```diff
@@ -12,29 +12,60 @@
...
         download_parser = parser.add_parser("env")
         download_parser.set_defaults(func=info_command_factory)
+        download_parser.add_argument(
+            "--accelerate-config_file",
+            default=None,
+            help="The accelerate config file to use for the default values in the launching script.",
+        )
+        download_parser.set_defaults(func=download_command_factory)
+
+    def __init__(self, accelerate_config_file, *args) -> None:
```
Comment (BenjaminBossan) [inline src/diffusers/commands/env.py:57]: Hmm, if we add a required argument here, why does `EnvironmentCommand()` inside of `info_command_factory` not fail?

File: src/diffusers/commands/env.py [hunk 7]
```diff
@@ -12,29 +12,60 @@
...
+        self._accelerate_config_file = accelerate_config_file
```
Comment (BenjaminBossan) [inline src/diffusers/commands/env.py:58]: Should this also call `super().__init__()`? Right now, this wouldn't do anything, but I think it's more cautious to call `super()` in child classes by default.

File: src/diffusers/commands/env.py [hunk 8]
```diff
@@ -43,33 +74,116 @@ def run(self):
...
         accelerate_version = "not installed"
+        accelerate_config = accelerate_config_str = "not found"
         if is_accelerate_available():
             import accelerate
+            from accelerate.commands.config import default_config_file, load_config_from_file
 
             accelerate_version = accelerate.__version__
+            # Get the default from the config file.
+            if self._accelerate_config_file is not None or os.path.isfile(default_config_file):
+                accelerate_config = load_config_from_file(self._accelerate_config_file).to_dict()
```
Comment (BenjaminBossan) [inline src/diffusers/commands/env.py:106]: Should this not load from `default_config_file` if `self._accelerate_config_file is None`?

File: src/diffusers/commands/env.py [hunk 9]
```diff
@@ -43,33 +74,116 @@ def run(self):
...
+
+                    start = out_str.find("VRAM (Total):")
+                    if start != -1:
+                        start += len("VRAM (Total):")
+                        end = out_str.find("\n", start)
+                        accelerator += " VRAM: " + out_str[start:end].strip()
+            except FileNotFoundError:
+                pass
+        else:
+            print("Are you crazy enough to use or build a new kind of OS like Terry A. Davis did?")
```
Comment (BenjaminBossan) [inline src/diffusers/commands/env.py:167]: It's funny but I'd prefer a more serious message here. Also, no shade at Terry Davis.

**Review by BenjaminBossan — COMMENTED** [review]
In general, LGTM, thanks for adding more information to the cli output. I have a couple of comments, please check.

Also, I'm not sure if the new additions copy some code from `huggingface-cli env` or `transformers-cli env` or if it's all written from scratch. If some of this is copied, it would be nice to add references.

File: src/diffusers/commands/env.py (hunk 9 above)
Comment (BenjaminBossan) [reply src/diffusers/commands/env.py]: I didn't think you meant that, but humor can be difficult with all the different contexts and languages, so I'd rather avoid it in these situations.

Comment (sayakpaul) [reply src/diffusers/commands/env.py]: Agree with Benjamin. Apologies that it slipped through my cracks.

Comment (standardAI) [reply src/diffusers/commands/env.py]: OK, that makes sense 👍.

**Comment by standardAI** [comment]
I added the references.
I actually wanted to think more about what else could be added. I remember I found one, but later I forgot :laughing:. Anyway, I will resolve the current conversations.

**Comment by DN6** [comment]
@standardAI Is this ready for review?

**Comment by standardAI** [comment]
Within this day, yes.

File: src/diffusers/commands/env.py (hunk 8 above)
Comment (standardAI) [reply src/diffusers/commands/env.py:107]: OK, makes sense. One should explicitly set `--accelerate-config_file` if one uses `accelerate` command; otherwise if one had arranged accelerate config file at its default location, but hasn't used `accelerate` command then accelerate config file is printed unnecessarily. Is this what you hightlight?

File: src/diffusers/commands/env.py [hunk 10]
```diff
@@ -43,33 +74,117 @@ def run(self):
...
         if is_accelerate_available():
             import accelerate
+            from accelerate.commands.config import default_config_file, load_config_from_file
 
             accelerate_version = accelerate.__version__
+            # Get the default from the config file.
+            # Taken from `transformers`.
+            if self._accelerate_config_file is not None or os.path.isfile(default_config_file):
+                accelerate_config = load_config_from_file(self._accelerate_config_file).to_dict()
+
```
Comment (standardAI) [inline src/diffusers/commands/env.py:108]: If someone wants to give `--accelerate-config_file` does (s)he want to give the default location of the config file? If so, would it be handy to be able to specified a small string?

File: src/diffusers/commands/env.py [hunk 11]
```diff
@@ -43,33 +76,121 @@ def run(self):
...
+            except FileNotFoundError:
+                pass
+        else:
+            print("It seems you are running an unusual OS. Could you fill in the accelerator manually?")
+
         info = {
-            "`diffusers` version": version,
-            "Platform": platform.platform(),
+            "Diffusers version": version,
+            "Location": os.path.dirname(os.path.abspath(version)).replace(getpass.getuser(), "USER"),
```
Comment (standardAI) [inline src/diffusers/commands/env.py:177]: I thought location may be usefull in terms of environment-related issues?

Comment (sayakpaul) [reply src/diffusers/commands/env.py:180]: Let's discard this. I don't think it provides any info and it might raise privacy concerns.

**Review by sayakpaul — APPROVED** [review]
This is looking quite nice to me. @asomoza would love a review from you too :)

**Comment by asomoza** [comment]
Nice work, I like all the info it gives, I have one issue with the accelerate config:

image

It doesn't detect the config file but I created it with the default command:

image

On the other side, even if it worked, this is what I have:

```
compute_environment: LOCAL_MACHINE
debug: false
distributed_type: 'NO'
downcast_bf16: 'no'
gpu_ids: all
machine_rank: 0
main_training_function: main
mixed_precision: 'no'
num_machines: 1
num_processes: 1
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

IMO it'll just clutter the copy & paste in the issues with something that's is not relevant most of the time, if the issue has something to do with accelerate we can ask for the config file but I don't think it's a good to have it all the time.

**Comment by DN6** [comment]
I think I agree with @asomoza.  Perhaps we exclude accelerate information from the CLI command for now. We can revisit adding it back in if there's a pressing need.

File: src/diffusers/commands/env.py (hunk 5 above)
Comment (standardAI) [reply src/diffusers/commands/env.py]: I think this might not be urgent. From now on, I want to focus on the theory.

**Comment by standardAI** [comment]
One needed to add `--accelerate-config_file` flag with the location of config_file. Anyway, I removed it.

**Comment by standardAI** [comment]
Thanks so much for all your invaluable feedback. This PR is ready for a final review now.

**Review by sayakpaul — APPROVED** [review]
Looking very nice.

**Review by BenjaminBossan — APPROVED** [review]
LGTM, thanks for extending the output with more useful information.

**Comment by standardAI** [comment]
Thanks for the approvals.
@DN6, WDYT?

**Comment by standardAI** [comment]
Thanks for the approvals.
Plus `bitsandbytes` and `quanto`?

**Comment by asomoza** [comment]
IMO `bitsandbytes` would be good.  For all the new models that use a `T5` text encoder, most people will need to use it in `8bit` or `4bit` so some issues can come from this library, especially in windows.

`quanto` in diffusers it's still an experiment so probably not yet.

**Comment by DN6** [comment]
@standardAI Can we do a final review and merge? I think this is ready.

**Comment by standardAI** [comment]
Yes, this is ready. I've only added `bitsandbytes` since the last approvals.

**Comment by standardAI** [comment]
The latest view:
```
❯ diffusers-cli env
An NVIDIA GPU may be present on this machine, but a CUDA-enabled jaxlib is not installed. Falling back to cpu.

Copy-and-paste the text below in your GitHub issue and FILL OUT the two last points.

- 🤗 Diffusers version: 0.28.0.dev0
- Platform: Ubuntu 24.04 LTS - Linux-6.8.0-31-generic-x86_64-with-glibc2.39
- Running on a notebook?: No
- Running on Google Colab?: No
- Python version: 3.10.14
- PyTorch version (GPU?): 2.3.0+cu121 (True)
- Flax version (CPU?/GPU?/TPU?): 0.8.3 (cpu)
- Jax version: 0.4.26
- JaxLib version: 0.4.26
- Huggingface_hub version: 0.23.0
- Transformers version: 4.40.2
- Accelerate version: 0.30.0
- PEFT version: 0.10.0
- Bitsandbytes version: 0.43.1
- Safetensors version: 0.4.3
- xFormers version: not installed
- Accelerator: NVIDIA GeForce GTX 1650, 4096 MiB VRAM
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>
```

**Comment by yiyixuxu** [comment]
@sayakpaul I think this is ready to merge now too, no?

**Review by sayakpaul — APPROVED** [review]
Thanks a bunch for working on this!

**Comment by standardAI** [comment]
Thanks for merging!
